### PR TITLE
Add support for a reference stable vector.

### PR DESCRIPTION
**Pull request by bbernhar**
Adds StableList, a new container type that holds any type of data contiguously when removal mostly occurs at the end.

### CODE REVIEW

**Comment by bbernhar** [comment]
@bjjones FYI. This is a STL-compatible container-only approach to storing linked-data contiguously (was looking to use it for non-residency cases). It's a similar idea to yours but keeps the container separate from allocation (like the STL), so we can swap them out independently.

File: src/gpgmm/utils/StableList.h [hunk 1]
```diff
@@ -0,0 +1,434 @@
+// Copyright 2021 The GPGMM Authors
```
Comment (bjjones) [inline src/gpgmm/utils/StableList.h:1]: nit: 2022

File: src/gpgmm/utils/StableList.h [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        */
+        void pop_back() {
+            erase(GetSizeWithUnused() - 1);
+        }
+
+        /** \brief Erase item by index.
+
+        The index is flatten or the value of size() - 1.
+
+        Erasing from a position other then the end will not erase immediately. Not until
```
Comment (bjjones) [inline src/gpgmm/utils/StableList.h:149]: nit: other than

File: src/gpgmm/utils/StableList.h [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+    A. The main reason to use StableList over a linked-list is to improve performance by using
+    contiguous memory for list-storage. A StableList will almost be always faster to iterate,
+    insert, and remove **unless** the majority of items reside in the middle of the list.
+
+    Q. How does StableList implementation differ from std::vector?
+
+    A. StableList is similar to STL vector except it is stable. Stable means pointers or
+    references to items are guarenteed to remain valid where removing any element does not
+    invalidate others. And unlike a std::vector, the cost of growth is predicable and constant.
```
Comment (bjjones) [inline src/gpgmm/utils/StableList.h:79]: typo: predicable

**Review by bjjones — APPROVED** [review]
LGTM 

Code seems fine. Doesn't look like erase()'d spots can be reused, which seems like a big drawback - could end up a lot of fragmentation, but I'm guessing you know that.

File: src/gpgmm/utils/StableList.h (hunk 2 above)
Comment (bbernhar) [reply src/gpgmm/utils/StableList.h:149]: Done.

File: src/gpgmm/utils/StableList.h (hunk 3 above)
Comment (bbernhar) [reply src/gpgmm/utils/StableList.h:79]: Done.

File: src/gpgmm/utils/StableList.h (hunk 1 above)
Comment (bbernhar) [reply src/gpgmm/utils/StableList.h:1]: Done.

**Comment by bbernhar** [comment]
> Doesn't look like erase()'d spots can be reused

Depends where they are. At the end, always reused. In the middle, not so. But I can lower the chunk size to limit waste (full chunks are reclaimable) or just discard those chunks. Not sure what approach is better.

**Comment by bbernhar** [comment]
Thanks!

**Pull request merged**